src/machine/machine_rp2040_adc.go

```diff
@@ -20,43 +39,112 @@ func InitADC() {
+}
+
+// The GetADCChannel method returns the channel associated with the ADC pin.
+// Returns !ok if the pin is not an ADC pin
+func (a ADC) GetADCChannel() (c ADCChannel, ok bool) {
+	ok = true
 	switch a.Pin {
-	case ADC0, ADC1, ADC2, ADC3:
-		a.Pin.Configure(PinConfig{Mode: PinAnalog})
+	case ADC0:
+		c = ADC0_CH
+	case ADC1:
+		c = ADC1_CH
+	case ADC2:
+		c = ADC2_CH
+	case ADC3:
+		c = ADC3_CH
 	default:
-		// invalid ADC
-		return
+		// Invalid ADC pin
+		ok = false
 	}
+	return c, ok
 }
 
-func (a ADC) Get() uint16 {
-	rp.ADC.CS.SetBits(uint32(a.getADCChannel()) << rp.ADC_CS_AINSEL_Pos)
+// Below methods are rp2040 specific, channel based access to the ADC peripheral.
+// The adcChannel methods provide access to the temperature sensor. As well as the pin based channels.
+
+// The Configure method sets the channel's associated pin to analog input mode.
+// Configuring the ADC_TEMP_CH channel will power on the temperature sensor.
+// The powered on temperature sensor increases ADC_AVDD current by approximately 40 μA.
+func (c ADCChannel) Configure(config ADCChConfig) {
+	if p, ok := c.Pin(); ok {
+		p.Configure(PinConfig{Mode: PinAnalog})
+	}
+	if c == ADC_TEMP_CH {
+		// Enable temperature sensor bias source
+		rp.ADC.CS.SetBits(rp.ADC_CS_TS_EN)
+	}
+}
+
+// The GetOnce method returns a one-shot ADC sample reading
+func (c ADCChannel) GetOnce() uint16 {
+	// Make it safe to sample multiple ADC channels in separate go routines.
+	adcLock.Lock()
+	rp.ADC.CS.ReplaceBits(uint32(c), 0b111, rp.ADC_CS_AINSEL_Pos)
 	rp.ADC.CS.SetBits(rp.ADC_CS_START_ONCE)
 
 	waitForReady()
+	adcLock.Unlock()
+
+	// rp2040 is a 12-bit ADC, scale raw reading to 16-bits.
+	return uint16(rp.ADC.RESULT.Get()) << 4
+}
 
-	// rp2040 uses 12-bit sampling, so scale to 16-bit
-	return uint16(rp.ADC.RESULT.Get() << 4)
+// The GetVoltage method does a one-shot sample and returns the reading as a voltage between 0.0 and 3.3V
+func (c ADCChannel) GetVoltage() (volts float32) {
+	return float32(c.GetOnce()>>4) * float32(3.3) / float32(1<<12)
 }
 
+// The GetTemp method does a one-shot sample of the internal temperature sensor.
+// Only works on the ADC_TEMP_CH channel. aka AINSEL=4. Other channels will return 0.0.
+func (c ADCChannel) GetTemp() (celsius float32) {
+	if c == ADC_TEMP_CH {
+		// T = 27 - (ADC_voltage - 0.706)/0.001721
+		return float32(27) - (c.GetVoltage()-float32(0.706))/float32(0.001721)
+	}
+	return 0.0
+}
+
+// The waitForReady function spins waiting for the ADC peripheral to become ready
 func waitForReady() {
 	for !rp.ADC.CS.HasBits(rp.ADC_CS_READY) {
 	}
 }
 
-func (a ADC) getADCChannel() uint8 {
-	switch a.Pin {
-	case ADC0:
-		return 0
-	case ADC1:
-		return 1
-	case ADC2:
-		return 2
-	case ADC3:
-		return 3
+// The Pin method returns the GPIO Pin associated with the ADC mux channel, if it has one.
+// Returns !ok if the channel does not have an associated GPIO, or if the channel number is invalid.
+func (c ADCChannel) Pin() (p Pin, ok bool) {
```

Comment:
Same as above, probably better to return an error? i.e:
`var errBadPinValue = errors.New("no associated pin with value")`